### PR TITLE
Dual-Pathway Cosmos3MoTDecoderLayer

### PR DESCRIPTION
**Pull request by parsley9877**
# Description

Adds the **Cosmos 3 Core Dual-Attention Operator (`CosmosDualAttention`)** and the **Dual-Pathway Mixture-of-Transformers Decoder Layer (`Cosmos3MoTDecoderLayer`)** in Flax NNX to MaxText.

### Context & Motivation
Cosmos 3 uses a Mixture-of-Transformers (MoT) architecture with two parallel transformer towers that maintain independent weights (projections, QK norms, pre/post RMSNorms, and MLPs) while sharing a single fused attention step:
- **Understanding pathway (`und`)**: processes text/prompt tokens using sample-isolated causal self-attention.
- **Generation pathway (`gen`)**: processes continuous video/multimodal latents using sample-isolated full bidirectional attention that also cross-attends to the understanding stream's keys and values.

### Implementation Details
1. **Dual-Pathway Packed Attention (`src/maxtext/layers/cosmos_attention.py`)**:
   - `CosmosPackingMetadata` & `build_cosmos_packing_metadata`: constructs cumulative sample boundary offsets (`causal_q_offsets`, `full_q_offsets`, `sample_kv_offsets`) and stream scatter/gather index maps (`packed_und_token_indexes`, `packed_gen_token_indexes`).
   - `unpack_streams` & `reinterleave_streams`: splits interleaved packed token sequences into understanding/generation streams and merges outputs back into global packed sequence order.
   - `compute_3d_mrope_cos_sin` & `apply_rotary_pos_emb`: implements 3D Multimodal Rotary Position Embeddings (3D M-RoPE) across temporal, height, and width sections.
   - `causal_understanding_attention` & `full_generative_attention`: supports `dot_product`, `flash`, and `splash` attention backends with strict cross-sample isolation masks (`build_causal_understanding_mask`, `build_full_generative_mask`, and Tokamax Splash mask compilation).
   - `CosmosDualAttention`: Flax NNX module encapsulating independent understanding and generation projections (`q_proj`/`k_proj`/`v_proj`/`o_proj` and `*_gen`), independent QK RMSNorms, and optional cross-pathway `k_norm_und_for_gen`.

2. **MoT Decoder Layer & Per-Tower MLPs (`src/maxtext/models/cosmos3.py`)**:
   - `Cosmos3MLP`: per-tower feed-forward block supporting both gated SwiGLU (`hidden_act="silu"`, Qwen3 backbone) and ungated squared-ReLU (`hidden_act="relu2"`, Nemotron backbone).
   - `Cosmos3MoTDecoderLayer`: composes `CosmosDualAttention` with independent understanding and generation towers:
     - Understanding pathway: `input_layernorm`, `to_q`, `to_k`, `to_v`, `to_out`, `norm_q`, `norm_k`, `post_attention_layernorm`, `mlp`.
     - Generation pathway: `input_layernorm_moe_gen`, `add_q_proj`, `add_k_proj`, `add_v_proj`, `to_add_out`, `norm_added_q`, `norm_added_k`, `post_attention_layernorm_moe_gen`, `mlp_moe_gen`.
   - Exposes `diffusers`-canonical parameter names (`to_q`, `add_q_proj`, `to_add_out`, `norm_added_q`, etc.) as zero-copy property aliases onto `self_attn` so checkpoint loaders and parity tests can address either naming convention without duplicating parameters in the NNX graph.

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Added two comprehensive unit and numerical-parity test suites:

1. **`tests/unit/cosmos_attention_test.py` (29 tests)**:
   - Stream packing/unpacking/re-interleaving round-trip tests and boundary offset checks.
   - Sample-isolation and causality invariance tests for both Kernel 1 (`causal_understanding_attention`) and Kernel 2 (`full_generative_attention`).
   - 3D M-RoPE shape and norm-preservation checks.
   - `jax.jit` compilation, `nnx.grad` backprop flow, and `dot_product` / `flash` / `splash` kernel parity.
   - **Golden GPU Parity (`CosmosDualAttentionGoldenParityTest`)**: verifies numerical parity against reference activations generated from NVIDIA's `cosmos-framework` on an A100 GPU (`gs://maxtext-test-assets/cosmos_attention_golden_data.npz`) across MHA, GQA, Nemotron asymmetric QK-norm, and no-RoPE configurations:

   ```text
   +-----------------------------------+---------------+---------------+----------+--------+
   | Case Name                         | Max Abs Error | Avg Abs Error | Atol Tol | Status |
   +-----------------------------------+---------------+---------------+----------+--------+
   | case_standard_dual_qk_norm        | 1.12e-2       | 8.30e-4       | 1.50e-2  |  PASS  |
   | case_nemotron_asymmetric_gqa      | 1.28e-2       | 9.10e-4       | 1.50e-2  |  PASS  |
   | case_standard_gqa                 | 9.40e-3       | 7.60e-4       | 1.50e-2  |  PASS  |
   | case_no_rope                      | 8.10e-3       | 6.50e-4       | 1.50e-2  |  PASS  |
   | case_block_aligned_* (dot/flash)  | <= 1.3e-2     | <= 9.2e-4     | 1.50e-2  |  PASS  |
   +-----------------------------------+---------------+---------------+----------+--------+
   ```

2. **`tests/unit/cosmos3_mot_layer_test.py` (19 tests)**:
   - Structural tests for `Cosmos3MLP` (`silu` vs `relu2`) and `Cosmos3MoTDecoderLayer` (weight-sharing aliases, causality, zero gen->und leakage, residual connections, `jax.jit`, and `nnx.grad` across both towers).
   - **Diffusers Golden Parity (`Cosmos3MoTDecoderLayerGoldenParityTest`)**: copies all weights from `diffusers.models.transformers.transformer_cosmos3.Cosmos3VLTextMoTDecoderLayer` (`und_len=16, gen_len=32, hidden_size=256`, stored in `gs://maxtext-test-assets/cosmos3_mot_golden_data.npz`) and verifies both `und_out` and `gen_out` against PyTorch in `float32` and `bfloat16`:

   ```text
   +-----------------+-----------+------------------+------------------+----------+--------+
   | Case Name       | Precision | und_out Max Diff | gen_out Max Diff | Atol Tol | Status |
   +-----------------+-----------+------------------+------------------+----------+--------+
   | mini_silu_mha   | float32   | 5.96e-7          | 4.77e-7          | 1.00e-4  |  PASS  |
   | mini_silu_gqa   | float32   | 7.15e-7          | 4.77e-7          | 1.00e-4  |  PASS  |
   | mini_relu2_asym | float32   | 7.15e-7          | 6.56e-7          | 1.00e-4  |  PASS  |
   | mini_silu_mha   | bfloat16  | 1.56e-2          | 1.56e-2          | 3.20e-2  |  PASS  |
   | mini_silu_gqa   | bfloat16  | 1.56e-2          | 1.56e-2          | 3.20e-2  |  PASS  |
   | mini_relu2_asym | bfloat16  | 1.56e-2          | 2.34e-2          | 3.20e-2  |  PASS  |
   +-----------------+-----------+------------------+------------------+----------+--------+
   ```

### Commands to Reproduce
```bash
pytest tests/unit/cosmos_attention_test.py -q
pytest tests/unit/cosmos3_mot_layer_test.py -q
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/m3/models/cosmos3/modeling_cosmos3.py [hunk 1]
```diff
@@ -0,0 +1,1578 @@
...
+from flax import nnx
+import jax
+from jax import sharding
+from jax.ad_checkpoint import checkpoint_name
+import jax.numpy as jnp
+from maxtext.kernels.attention import jax_flash_attention
+from maxtext.kernels.tokamax_splash_attention import splash_attention_kernel
+from maxtext.kernels.tokamax_splash_attention import splash_attention_mask
+from maxtext.kernels.tokamax_splash_attention import splash_attention_mask_info
+import numpy as np
```
Comment (hengtaoguo) [inline src/maxtext/m3/models/cosmos3/modeling_cosmos3.py:74]: Gladly this file doesn't have dependency on previous attention layers!

File: tests/unit/cosmos3_mot_layer_test.py
Comment (hengtaoguo) [inline tests/unit/cosmos3_mot_layer_test.py:1]: I would prefer to combine these two tests into a single one `cosmos3_layers_test.py` for consolidation.

https://screenshot-v2.corp.google.com/1908un667ic9o

File: src/maxtext/m3/models/cosmos3/modeling_cosmos3.py
Comment (hengtaoguo) [inline src/maxtext/m3/models/cosmos3/modeling_cosmos3.py:1]: So far, these are all standalone building blocks right?

File: src/maxtext/m3/models/cosmos3/modeling_cosmos3.py [hunk 2]
```diff
@@ -0,0 +1,1578 @@
...
+    return self.self_attn.k_norm_und_for_gen
+
+  def __call__(
+      self,
+      und_seq: jax.Array,
+      gen_seq: jax.Array,
+      metadata: CosmosPackingMetadata,
+      cos: jax.Array | None = None,
+      sin: jax.Array | None = None,
+  ) -> tuple[jax.Array, jax.Array]:
```
Comment (hengtaoguo) [inline src/maxtext/m3/models/cosmos3/modeling_cosmos3.py:1583]: Can we add a concise doc string to indicate the in-and-out tensor shape/layout? This also applies to other blocks such as MLP.

**Review by hengtaoguo — COMMENTED** [review]
Thanks for the quick adaptation! cc @bvandermoon I'm not sure M3 is ready to accept codes addition now. Would you have any suggestions?

AFAIK, this PR is pretty isolated with only building blocks and associated tests.

File: src/maxtext/m3/models/cosmos3/modeling_cosmos3.py
Comment (parsley9877) [reply src/maxtext/m3/models/cosmos3/modeling_cosmos3.py:1]: Yes. Attention is implemented as a standalone block and is tested separately, then it is used in the full decoder block, and decoder block is also tested separately. The test files will be merged per your other comment.

File: src/maxtext/m3/models/cosmos3/modeling_cosmos3.py (hunk 2 above)
Comment (parsley9877) [reply src/maxtext/m3/models/cosmos3/modeling_cosmos3.py:1583]: Sure. I added a docstring for every __call__ method. I also added a general diagram of block dataflow on top of the file.

File: tests/unit/cosmos3_mot_layer_test.py
Comment (parsley9877) [reply tests/unit/cosmos3_mot_layer_test.py:1]: Done.

File: src/maxtext/models/cosmos3/cosmos3_sharding.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+  """Maps semantic logical tensor axes to explicit mesh PartitionSpecs."""
+
+  @abc.abstractmethod
+  def map_axis(
+      self,
+      axis: str,
+      tensor_name: str,
+      tensor_type: TensorType,
+  ) -> MeshAxisMapping:
+    """Maps one logical axis to physical mesh axis name(s), or None."""
```
Comment (bvandermoon) [inline src/maxtext/models/cosmos3/cosmos3_sharding.py]: I am working on adding the generic sharding support currently. Could we separate this file into a separate follow-up PR so we can rely on that once it's merged?

Comment (parsley9877) [reply src/maxtext/models/cosmos3/cosmos3_sharding.py]: Sounds good!  removed the sharding file.

File: tests/unit/cosmos3_layers_test.py
Comment (hengtaoguo) [inline tests/unit/cosmos3_layers_test.py:1]: Could you mark this test as cpu_only and scheduled_only ([example](https://github.com/AI-Hypercomputer/maxtext/pull/4633))?

It's nice to have tests but important to maintain CI speed by limiting the number of valuable tests. Could you help check the enabled tests to see which of them are the most critical ones? Thanks!

File: src/maxtext/m3/models/cosmos3/modeling_cosmos3.py
Comment (hengtaoguo) [inline src/maxtext/m3/models/cosmos3/modeling_cosmos3.py:1]: This is great. Let's move the folder cosmos3 to `src/maxtext/m3/models/cosmos3`, and rename `cosmos3.py` to `modeling_cosmos3.py`.

cc @bvandermoon @lydhr

**Review by hengtaoguo — APPROVED** [review]

File: src/maxtext/m3/models/cosmos3/modeling_cosmos3.py
Comment (parsley9877) [reply src/maxtext/m3/models/cosmos3/modeling_cosmos3.py:1]: Done! Moved the implementation to src/maxtext/m3/models/cosmos3/ and renamed cosmos3.py to modeling_cosmos3.py.

File: tests/unit/cosmos3_layers_test.py
Comment (parsley9877) [reply tests/unit/cosmos3_layers_test.py:1]: Thanks! We split the tests like this:

Presubmit (smoke suite): 30 fast tests.
Scheduled CI (@pytest.mark.scheduled_only): 21 heavier tests, including full JIT compilation & autodiff gradient tracing, as well as the complete numerical golden parity suites (12 cases against GPU cosmos-framework and 6 cases against diffusers).

File: src/maxtext/m3/models/cosmos3/modeling_cosmos3.py [hunk 4]
```diff
@@ -0,0 +1,1630 @@
...
+
+from flax import nnx
+import jax
+from jax import sharding
+from jax.ad_checkpoint import checkpoint_name
+import jax.numpy as jnp
+from maxtext.kernels.attention import jax_flash_attention
+from maxtext.kernels.tokamax_splash_attention import splash_attention_kernel
+from maxtext.kernels.tokamax_splash_attention import splash_attention_mask
+from maxtext.kernels.tokamax_splash_attention import splash_attention_mask_info
```
Comment (bvandermoon) [inline src/maxtext/m3/models/cosmos3/modeling_cosmos3.py]: Please remove the kernel support for now. It's not handled in m3 yet and shouldn't be part of the model file

File: src/maxtext/m3/models/cosmos3/modeling_cosmos3.py [hunk 5]
```diff
@@ -0,0 +1,1630 @@
...
+def _make_linear(
+    in_features: int,
+    out_features: int,
+    *,
+    use_bias: bool = False,
+    dtype: Any = jnp.float32,
+    weight_dtype: Any = jnp.float32,
+    kernel_init: Callable[..., jax.Array] | None = None,
+    kernel_axes: tuple[None | str, ...] = (),
+    sharding_hook: Callable[[Any, str], Any] | None = None,
```
Comment (bvandermoon) [inline src/maxtext/m3/models/cosmos3/modeling_cosmos3.py]: Please remove the sharding logic for now. t's not handled in m3 yet and shouldn't be part of the model file

File: src/maxtext/m3/models/cosmos3/modeling_cosmos3.py [hunk 6]
```diff
@@ -0,0 +1,1630 @@
...
+        sin_gen = sin[metadata.packed_gen_token_indexes]
+        q_gen = apply_rotary_pos_emb(q_gen, cos_gen, sin_gen, unsqueeze_dim=1)
+        k_gen = apply_rotary_pos_emb(k_gen, cos_gen, sin_gen, unsqueeze_dim=1)
+
+    q_und = checkpoint_name(hook(q_und, "query"), "query_proj")
+    k_und = checkpoint_name(hook(k_und, "key"), "key_proj")
+    v_und = checkpoint_name(hook(v_und, "value"), "value_proj")
+    q_gen = checkpoint_name(hook(q_gen, "gen_query"), "gen_query_proj")
+    k_gen = checkpoint_name(hook(k_gen, "gen_key"), "gen_key_proj")
+    v_gen = checkpoint_name(hook(v_gen, "gen_value"), "gen_value_proj")
```
Comment (bvandermoon) [inline src/maxtext/m3/models/cosmos3/modeling_cosmos3.py]: Not sure if we want these `checkpoint_name` pieces added in m3 yet

**Review by bvandermoon — COMMENTED** [review]
Thanks for the PR @parsley9877. We should talk more about this offline. The goal of m3 is to use a single-file-per-model type of setup. This will help with readability. But the m3 work has just started and virtually all features are missing because of that. 

This PR takes the pattern of moving the model logic to a single file. But it seems like it is trying to make up for the features that aren't available in m3 yet (like kernel support, shardings, etc.) by fitting them all into this model file as well. These types of features will be shared for all of m3 rather than fitting them into the model files.

If we are going to support this model in m3 right now, then we will want to fit the m3 pattern of keeping just the core model logic in this file, and get access to performance features as they become available for m3 as a whole. Otherwise, we are sacrificing the goals of m3 and risking setting precedent for other models. Let me know your thoughts on this

**Comment by lydhr** [comment]
> Thanks for the PR @parsley9877. We should talk more about this offline. The goal of m3 is to use a single-file-per-model type of setup. This will help with readability. But the m3 work has just started and virtually all features are missing because of that.
> 
> This PR takes the pattern of moving the model logic to a single file. But it seems like it is trying to make up for the features that aren't available in m3 yet (like kernel support, shardings, etc.) by fitting them all into this model file as well. These types of features will be shared for all of m3 rather than fitting them into the model files.
> 
> If we are going to support this model in m3 right now, then we will want to fit the m3 pattern of keeping just the core model logic in this file, and get access to performance features as they become available for m3 as a whole. Otherwise, we are sacrificing the goals of m3 and risking setting precedent for other models. Let me know your thoughts on this

+1 FYI. We have 2 pending M3 models' PRs in progress, aligning with our framework TL's review. Will keep you posted once they settle down!
